drivers/mm/mm_drv_intel_adsp_mtl_tlb.c

```diff
@@ -401,6 +401,80 @@ int sys_mm_drv_unmap_region(void *virt, size_t size)
 	return sys_mm_drv_simple_unmap_region(va, size);
 }
 
+int sys_mm_drv_update_page_flags(void *virt, uint32_t flags)
+{
+	k_spinlock_key_t key;
+	uint32_t entry_idx;
+	uint16_t entry;
+	uint16_t *tlb_entries = UINT_TO_POINTER(TLB_BASE);
+	int ret = 0;
+
+	/* Use cached virtual address */
+	uintptr_t va = POINTER_TO_UINT(z_soc_cached_ptr(virt));
+
+	/* Make sure inputs are page-aligned and check bounds of virtual address space */
+	CHECKIF(!sys_mm_drv_is_addr_aligned(va) ||
+		(va < UNUSED_L2_START_ALIGNED) ||
+		(va >= (CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE))) {
+		return -EINVAL;
+	}
+
+	key = k_spin_lock(&tlb_lock);
+
+	entry_idx = get_tlb_entry_idx(va);
+
+	entry = tlb_entries[entry_idx];
+
+	/* Check entry is already mapped */
+	if (!(entry & TLB_ENABLE_BIT)) {
+		ret = -EFAULT;
+		goto out;
+	}
+
+	/* Clear the access flags */
+	entry &= ~(TLB_EXEC_BIT | TLB_WRITE_BIT);
+
+	/* Set new permissions for this entry */
+	entry |= flags_to_tlb_perms(flags);
+
+	tlb_entries[entry_idx] = entry;
+
+out:
+	k_spin_unlock(&tlb_lock, key);
+	return ret;
+}
+
+int sys_mm_drv_update_region_flags(void *virt, size_t size, uint32_t flags)
+{
+	k_spinlock_key_t key;
+	int ret = 0;
+	size_t offset;
+	uint8_t *va;
+
+	CHECKIF(!sys_mm_drv_is_virt_addr_aligned(virt) ||
+		!sys_mm_drv_is_size_aligned(size)) {
+		return -EINVAL;
+	}
+
+	va = (__sparse_force uint8_t *)z_soc_cached_ptr(virt);
+
+	key = k_spin_lock(&sys_mm_drv_common_lock);
+
+	for (offset = 0; offset < size; offset += CONFIG_MM_DRV_PAGE_SIZE) {
+		int ret2 = sys_mm_drv_update_page_flags(va, flags);
+
+		if (ret2 != 0) {
+			__ASSERT(false, "cannot update page flags %p to %x\n", va, flags);
+
+			ret = ret2;
+		}
+		va += CONFIG_MM_DRV_PAGE_SIZE;
+	}
+
+	k_spin_unlock(&sys_mm_drv_common_lock, key);
+	return ret;
+}
```

Comment:
This looks very much like the common implementation - wouldn't using it, like https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/mm/mm_drv_intel_adsp_mtl_tlb.c#L327 be simpler?

Reply:
@edersondisouza I'm sorry, I dont understand. How spinlock is related with `(__sparse_force void *)z_soc_cached_ptr(virt);`?

Reply:
Oh, sorry - it's not about the specific line, but that implementation of the function, that basically reuses the one on  `mm_drv_common.c`- `sys_mm_drv_simple_map_array` - can't something like that be done here, with `sys_mm_drv_simple_update_region_flags`? The code seems very similar, it would be nice if it's possible to reuse it.

Reply:
any updates?